doit/dependency.py

```diff
@@ -693,6 +693,12 @@ def get_status(self, task, tasks_dict, get_log=False):
 class UptodateCalculator(object):
     """Base class for 'uptodate' that need access to all tasks
     """
+    
```

Comment:
Pickle debug. get_val is a bound method of DbmDB which ultimately leads to the pickling of an instance of DbmDB and its attribute DB which isn't picklable.
```
pickling object [(<doit.task.result_dep object at 0x042AFCF0>, [], {})] of type <type 'list'>
pickling object (<doit.task.result_dep object at 0x042AFCF0>, [], {}) of type <type 'tuple'>
pickling object <doit.task.result_dep object at 0x042AFCF0> of type <class 'doit.task.result_dep'>
pickling object <class 'doit.task.result_dep'> of type <type 'type'>
pickling object () of type <type 'tuple'>
pickling object {'result_name': '_result:t2', 'dep_name': 't2', 'tasks_dict': {'t2': <Task: t2>, 't1': <Task: t1>}, 'get_val': <bound method DbmDB.wrap of <doit.dependency.DbmDB object at 0x042AF810>>} of type <type 'dict'>
pickling object result_name of type <type 'str'>
pickling object _result:t2 of type <type 'str'>
pickling object dep_name of type <type 'str'>
pickling object t2 of type <type 'str'>
pickling object tasks_dict of type <type 'str'>
pickling object {'t2': <Task: t2>, 't1': <Task: t1>} of type <type 'dict'>
pickling object get_val of type <type 'str'>
pickling object <bound method DbmDB.wrap of <doit.dependency.DbmDB object at 0x042AF810>> of type <type 'instancemethod'>
pickling object <built-in function getattr> of type <type 'builtin_function_or_method'>
pickling object (<doit.dependency.DbmDB object at 0x042AF810>, 'wrap') of type <type 'tuple'>
pickling object <doit.dependency.DbmDB object at 0x042AF810> of type <class 'doit.dependency.DbmDB'>
pickling object <class 'doit.dependency.DbmDB'> of type <type 'type'>
pickling object () of type <type 'tuple'>
pickling object {'_dbm': {}, 'name': 'c:\\\\users\\\\gds\\\\appdata\\\\local\\\\temp\\\\pytest-25\\\\test_getargs_RunnerClass2_0\\\\testdb', '_db': {}, 'dirty': set([])} of type <type 'dict'>
pickling object _dbm of type <type 'str'>
pickling object {} of type <class 'bsddb._DBWithCursor'>
pickling object <class 'bsddb._DBWithCursor'> of type <class 'abc.ABCMeta'>
pickling object () of type <type 'tuple'>
pickling object {'saved_dbc_key': None, 'db': <DB object at 0x042BD930>, 'dbc': None, '_kill_iteration': False, '_in_iter': 0, '_cursor_refs': {}} of type <type 'dict'>
pickling object saved_dbc_key of type <type 'str'>
pickling object None of type <type 'NoneType'>
pickling object db of type <type 'str'>
pickling object <DB object at 0x042BD930> of type <type 'DB'>
pickling object <type 'DB'> of type <type 'type'>
error. skipping...
```
Trace from AppVeyor:
```
================================== FAILURES =================================== 
_______________ TestRunner_run_tasks.test_getargs[RunnerClass2] _______________ 
 
self = <tests.test_runner.TestRunner_run_tasks object at 0x023518F0> 
reporter = <tests.test_runner.FakeReporter object at 0x02351F50> 
RunnerClass = <class 'doit.runner.MRunner'>
dep_manager = <doit.dependency.Dependency object at 0x02D27ED0> 
 
    def test_getargs(self, reporter, RunnerClass, dep_manager): 
        t1 = Task("t1", [(use_args,)], getargs=dict(arg1=('t2','myarg')) ) 
        t2 = Task("t2", [(make_args,)]) 
        my_runner = RunnerClass(dep_manager, reporter) 
>       my_runner.run_tasks(TaskDispatcher({'t1':t1, 't2':t2}, [], ['t1', 't2'])) 
 
tests\test_runner.py:488:  
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
doit\runner.py:454: in run_tasks 
    proc_list = self._run_start_processes(job_q, result_q) 
doit\runner.py:427: in _run_start_processes 
    process.start() 
C:\Python27\lib\multiprocessing\process.py:130: in start 
    self._popen = Popen(self) 
C:\Python27\lib\multiprocessing\forking.py:277: in __init__ 
    dump(process_obj, to_child, HIGHEST_PROTOCOL) 
C:\Python27\lib\multiprocessing\forking.py:199: in dump 
    ForkingPickler(file, protocol).dump(obj) 
C:\Python27\lib\pickle.py:224: in dump 
    self.save(obj) 
C:\Python27\lib\pickle.py:331: in save 
    self.save_reduce(obj=obj, *rv) 
C:\Python27\lib\pickle.py:419: in save_reduce 
    save(state) 
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:649: in save_dict 
    self._batch_setitems(obj.iteritems()) 
C:\Python27\lib\pickle.py:681: in _batch_setitems 
    save(v) 
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\multiprocessing\forking.py:67: in dispatcher 
    self.save_reduce(obj=obj, *rv)
C:\Python27\lib\pickle.py:401: in save_reduce
    save(args) 
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self
C:\Python27\lib\pickle.py:548: in save_tuple 
    save(element) 
C:\Python27\lib\pickle.py:331: in save 
    self.save_reduce(obj=obj, *rv)
C:\Python27\lib\pickle.py:419: in save_reduce 
    save(state) 
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:649: in save_dict 
    self._batch_setitems(obj.iteritems()) 
C:\Python27\lib\pickle.py:681: in _batch_setitems 
    save(v) 
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:649: in save_dict 
    self._batch_setitems(obj.iteritems())
C:\Python27\lib\pickle.py:681: in _batch_setitems 
    save(v)
C:\Python27\lib\pickle.py:331: in save 
    self.save_reduce(obj=obj, *rv) 
C:\Python27\lib\pickle.py:419: in save_reduce 
    save(state)
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:649: in save_dict 
    self._batch_setitems(obj.iteritems()) 
C:\Python27\lib\pickle.py:681: in _batch_setitems 
    save(v)
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:600: in save_list 
    self._batch_appends(iter(obj)) 
C:\Python27\lib\pickle.py:636: in _batch_appends 
    save(tmp[0])
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:548: in save_tuple 
    save(element) 
C:\Python27\lib\pickle.py:331: in save 
    self.save_reduce(obj=obj, *rv) 
C:\Python27\lib\pickle.py:419: in save_reduce 
    save(state)
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:649: in save_dict 
    self._batch_setitems(obj.iteritems()) 
C:\Python27\lib\pickle.py:681: in _batch_setitems 
    save(v)
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\multiprocessing\forking.py:67: in dispatcher 
    self.save_reduce(obj=obj, *rv)
C:\Python27\lib\pickle.py:401: in save_reduce 
    save(args) 
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self
C:\Python27\lib\pickle.py:548: in save_tuple 
    save(element) 
C:\Python27\lib\pickle.py:331: in save 
    self.save_reduce(obj=obj, *rv) 
C:\Python27\lib\pickle.py:419: in save_reduce 
    save(state) 
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:649: in save_dict 
    self._batch_setitems(obj.iteritems())
C:\Python27\lib\pickle.py:681: in _batch_setitems 
    save(v) 
C:\Python27\lib\pickle.py:331: in save 
    self.save_reduce(obj=obj, *rv) 
C:\Python27\lib\pickle.py:419: in save_reduce 
    save(state)
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
C:\Python27\lib\pickle.py:649: in save_dict 
    self._batch_setitems(obj.iteritems())
C:\Python27\lib\pickle.py:681: in _batch_setitems 
    save(v) 
C:\Python27\lib\pickle.py:331: in save 
    self.save_reduce(obj=obj, *rv)
C:\Python27\lib\pickle.py:396: in save_reduce
    save(cls)
C:\Python27\lib\pickle.py:286: in save 
    f(self, obj) # Call unbound method with explicit self 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
 
self = <multiprocessing.forking.ForkingPickler instance at 0x02D7B198> 
obj = <type 'DB'>, name = 'DB', pack = <built-in function pack>
 
    def save_global(self, obj, name=None, pack=struct.pack): 
        write = self.write 
        memo = self.memo 
     
        if name is None: 
            name = obj.__name__ 
     
        module = getattr(obj, "__module__", None) 
        if module is None: 
            module = whichmodule(obj, name) 
     
        try: 
            __import__(module) 
            mod = sys.modules[module] 
            klass = getattr(mod, name) 
        except (ImportError, KeyError, AttributeError): 
            raise PicklingError( 
                "Can't pickle %r: it's not found as %s.%s" % 
>               (obj, module, name)) 
E           PicklingError: Can't pickle <type 'DB'>: it's not found as __builtin__.DB 
 
C:\Python27\lib\pickle.py:748: PicklingError
============== 1 failed, 690 passed, 19 skipped in 14.81 seconds ============== 
Command exited with code 1
```